src/main/java/io/quarkus/gizmo/BytecodeCreator.java

```diff
@@ -267,21 +269,68 @@ default ResultHandle load(Enum<?> val) {
     /**
      * Returns a {@link ResultHandle} representing the specified class
      *
+     * This will be loaded from the TCCL, rather than via bytecode reference.
+     *
      * @param val The class to load
      * @return A {@link ResultHandle} representing the specified class
      */
     default ResultHandle loadClass(Class<?> val) {
-        return loadClass(val.getName());
+        return loadClass(val.getName(),true);
```

Comment:
I'm not entirely sure the TCCL should be used by default. It's probably OK in most cases but anyway is ok for all cases? I'm talking about backward compatibility.

Reply:
If there are cases in Quarkus where this causes problems we can change them to use the explicit version, however I don't think this will be the case. It's much more likely that this change will fix bugs we have not discovered yet.

Reply:
It might fix some issues but it just does not feel OK to me. What was previously a simple `ldc` from the constant pool would become `Thread.currentThread().getContextClassLoader()` + `Class.forName()` (with TCCL cached in the current scope IIUIC) and there's no way to detect this change (except if your code breaks). But maybe I'm just "hypersensitive" ;-)

@geoand @gsmet @Ladicek @Sanne and others WDYT?

Reply:
I love the idea of fixing the problem once and for all, but I share Martin's concern here, that `loadClass` should be `ldc` by default.

Reply:
Isn't it (generally) expected that all code gets load from the contextual classloader? That's how a normal Java program works.

I think this patch makes sense as a default; I also share some nervousness about changing such a core detail but let's see CI: failures might indeed indicate some need for some code to not use the default and we can address that, this still makes sense conceptually.

Reply:
It is not how normal Java programs work. In a normal situation, when class A refers to class B (through normal linguistic means, not through reflection), then B is loaded by the same classloader that loaded A. TCCL doesn't even have to be set (and `Class.forName()` apparently uses the bootstrap CL when passed `null`).

Now, it is rather common to prefer loading classes from TCCL in applications that are "containers" (be it application servers, or Quarkus). If Gizmo serves Quarkus and doesn't strive to be generic, then this change probably makes sense. I do have some mixed feelings about it too, but can't really articulate those.

Reply:
> It is not how normal Java programs work. In a normal situation, when class A refers to class B (through normal linguistic means, not through reflection), then B is loaded by the same classloader that loaded A. TCCL doesn't even have to be set (and Class.forName() apparently uses the bootstrap CL when passed null).

+1

Reply:
doh right. Sorry - best not try to rush emails before coffee :)

still, what I said is generally the expectation in Quarkus specifically, as the contect classloader is _the_ runtime classloader and delegating to it should be the safe choice, as it then is able to forward to the right classloaders, and take into account dev/prod differences.

Reply:
I will be incredibly surprised if this causes any issues in Quarkus, and if it does it means there is something very wrong going on that should be fixed anyway. 

Given that people seem to be worried about changing this I could just add a new method and change every call in Quarkus over to use the new one, and blacklist the current method.